File: C++/isMatchSimple.cpp
/**
 * @File    : isMatchSimple.cpp
 * @Brief   : 通配符匹配
 * @Link    : https://leetcode-cn.com/problems/wildcard-matching/
 * @Author  : Wei Li
 * @Date    : 2021-08-17
*/

#include <iostream>
#include <string>

class Solution
{
public:
    bool isMatch(std::string s, std::string p)
    {
        auto allStars = [](const std::string &str, int left, int right)
        {
            for (int i = left; i < right; ++i)
            {
                if (str[i] != '*')
                {
                    return false;
                }
            }
            return true;
        };
        auto charMatch = [](char u, char v)
        {
            return u == v || v == '?';
        };

        while (s.size() && p.size() && p.back() != '*')
        {
            if (charMatch(s.back(), p.back()))
            {
                s.pop_back();
                p.pop_back();
            }
            else
            {
                return false;
            }
        }
        if (p.empty())
        {
            return s.empty();
        }

        int sIndex = 0, pIndex = 0;
        int sRecord = -1, pRecord = -1;
        while (sIndex < s.size() && pIndex < p.size())
        {
            if (p[pIndex] == '*')
            {
                ++pIndex;
                sRecord = sIndex;
                pRecord = pIndex;
            }
            else if (charMatch(s[sIndex], p[pIndex]))
            {
                ++sIndex;
                ++pIndex;
            }
            else if (sRecord != -1 && sRecord + 1 < s.size())
            {
                ++sRecord;
                sIndex = sRecord;
                pIndex = pRecord;
            }
            else
            {
                return false;
            }
        }
        return allStars(p, pIndex, p.size());
    }
};

// ----------------------------
int main(int argc, char **argv)
{
    // std::string str = "adceb";
    // std::string pattern = "*a*b";
    std::string str = "acdcb";
    std::string pattern = "a*c?b";

    Solution solution;
    bool is_match_simple = solution.isMatch(str, pattern);
    std::cout << "The solution of this problem is " << std::boolalpha << is_match_simple << std::endl;

    return 0;
}

File: Python/isMatchSimple.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: isMatchSimple.py
@Function: 通配符匹配
@Link: https://leetcode-cn.com/problems/wildcard-matching/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-17
"""


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        def allStars(st: str, left: int, right: int) -> bool:
            return all(st[i] == '*' for i in range(left, right))

        def charMatch(u: str, v: str) -> bool:
            return u == v or v == '?'

        sRight, pRight = len(s), len(p)
        while sRight > 0 and pRight > 0 and p[pRight - 1] != '*':
            if charMatch(s[sRight - 1], p[pRight - 1]):
                sRight -= 1
                pRight -= 1
            else:
                return False

        if pRight == 0:
            return sRight == 0

        sIndex, pIndex = 0, 0
        sRecord, pRecord = -1, -1
        while sIndex < sRight and pIndex < pRight:
            if p[pIndex] == '*':
                pIndex += 1
                sRecord, pRecord = sIndex, pIndex
            elif charMatch(s[sIndex], p[pIndex]):
                sIndex += 1
                pIndex += 1
            elif sRecord != -1 and sRecord + 1 < sRight:
                sRecord += 1
                sIndex, pIndex = sRecord, pRecord
            else:
                return False

        return allStars(p, pIndex, pRight)


# -------------------------
if __name__ == "__main__":
    # s = "acdcb"
    # p = "a*c?b"

    s = "adceb"
    p = "*a*b"

    solution = Solution()
    is_match_simple = solution.isMatch(s, p)
    print(f"The solution of this problem is {is_match_simple}")
